Differentiate alt and dummy cycles in QSPIF

kyle-cypress opened the pull request:
### Description
Proposed fix for #11295 

Propagate separate alt cycle and dummy cycle counts from QSPIFBlockDevice
down to the qspi driver, so that drivers which handle the two separately have
enough information to do so.

Submitting the pull request now to obtain feedback on the change. Greentea test logs will be attached soon.

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers
@ARMmbed/team-cypress
<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

Commit:
Differentiate alt and dummy cycles in QSPIF

Propagate separate alt cycle and dummy cycle counts from QSPIFBlockDevice
down to the qspi driver, so that drivers which handle the two separately have
enough information to do so.

VeijoPesonen reviewed (approved):
Looks good to me. But I would like to see someone else also to review this as I'm not that familiar with the subject yet.

bulislaw reviewed (commented):
@davidsaada @bentcooke can you guys have a look please

Commit:
Allow for arbitrary QSPI alt sizes

The QSPI spec allows alt to be any size that is a multiple of the
number of data lines. For example, Micron's N25Q128A uses only a
single alt cycle for all read modes (1, 2, or 4 bits depending on
how many data lines are in use).

maciejbocianski commented on targets/TARGET_STM/qspi_api.c:
```diff
@@ -249,10 +278,40 @@ void qspi_prepare_command(const qspi_command_t *command, QSPI_CommandTypeDef *st
+            // Align alt value to the end of the most significant byte
+            st_command->AlternateBytes = command->alt.value << leftover_bits;
+        } else {
+            rounded_size -= 1;
```
I think this line shouldn't be here. It cause that `rounded_size` falls below 0 in some cases

kyle-cypress replied:
I agree. Addressed in 9b32c0f pushed to #11602 , and rebased this PR on #11602 again

maciejbocianski reviewed (changes requested):
I'm getting quite different results comparing to @jeromecoutant results.
`DISCO_F413ZH` passes with and without fix while 'DISCO_L475VG_IOT01A' is failing with fix.

Commit:
Fix possible negative QSPI alt count on STM

Remove an extraneous decrement operation in cases where the alt
bits size is a multiple of 8.

jeromecoutant reviewed (approved):
Tested with ARMC6 with:
- DISCO_L496AG
- DISCO_L476VG
- DISCO_F769NI
- DISCO_F469NI
- DISCO_F746NG
- DISCO_F413ZH
- DISCO_L475VG_IOT01A

@LMESTM @MarceloSalazar
